cli/src/lib.rs

```diff
@@ -135,39 +136,26 @@ pub trait CommandT {
 /// Error returned by [CommandT::run].
 ///
 /// Implements [From] for client errors.
-#[derive(Debug, derive_more::From)]
+#[derive(Debug, derive_more::From, ThisError)]
 pub enum CommandError {
+    #[error("Client error: {0}")]
     ClientError(Error),
+
+    #[error("Transaction {tx_hash:?} failed in block {block_hash:?}")]
     FailedTransaction {
         tx_hash: TxHash,
         block_hash: BlockHash,
     },
-    OrgNotFound {
-        org_id: OrgId,
-    },
+
+    #[error("Cannot find org {org_id:?}")]
+    OrgNotFound { org_id: OrgId },
+
+    #[error("Cannot find project {project_name:?}.{org_id:?}")]
     ProjectNotFound {
         project_name: ProjectName,
         org_id: OrgId,
     },
-    AccountStorageError(account_storage::Error),
-}
 
-impl core::fmt::Display for CommandError {
-    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
-        match self {
-            CommandError::ClientError(error) => write!(f, "Client error: {}", error),
-            CommandError::FailedTransaction {
-                tx_hash,
-                block_hash,
-            } => write!(f, "Transaction {} failed in block {}", tx_hash, block_hash),
-            CommandError::OrgNotFound { org_id } => write!(f, "Cannot find org {}", org_id),
-            CommandError::ProjectNotFound {
-                project_name,
-                org_id,
-            } => write!(f, "Cannot find project {}.{}", project_name, org_id),
-            CommandError::AccountStorageError(error) => {
-                write!(f, "Account storage error: {}", error)
-            }
-        }
-    }
+    #[error("Account storage error: {0}")]
```

Comment:
This now gives an error like this
~~~
ERROR: Account storage error: A serde json error occured: invalid length 31, expected an array of length 32 at line 3 column 75
~~~
This is not super readable. I’d suggest 
```suggestion
    #[error("{0}")]
```
and then adjust the message of the account error so that they are understandable to the user. Ideally the message would be something like this
~~~
ERROR: Failed to read accounts file: invalid length 31, expected an array of length 32 at line 3 column 75
~~~

Reply:
Ok, sounds good :)

Given that IOErrors and Json errors can happen on read/write or deserialize/serialize, I will add corresponding variants to `account_storage::Error` in order to provide that more detailed kind of error message.

Reply:
@geigerzaehler could you have a look? I added `FailedWriting` and `FailedRead` variants that wrap themselves variants for IO and json (de)serialization use cases. We don't benefit from having this extra sub error type but how else would we have a generic error type on the high-level variants?

I.e, Now we have:

``` rust
#[error("Failed to write to the accounts file: {0}")]
FailedWrite(WritingError),

 #[error("Failed to read the accounts file: {0}")]
 FailedRead(ReadingError),
```

where

``` rust
#[derive(Debug, ThisError)]
pub enum WritingError {
    #[error("{0}")]
    IO(IOError),

    #[error("{0}")]
    Serialization(serde_json::Error),
}

/// Possible errors when reading the accounts file.
#[derive(Debug, ThisError)]
pub enum ReadingError {
    #[error("{0}")]
    IO(IOError),

    #[error("{0}")]
    Deserialization(serde_json::Error),
}
```

whereas I'd prefer to just have:

``` rust
#[error("Failed to write to the accounts file: {0}")]
FailedWrite(AnyError),

 #[error("Failed to read the accounts file: {0}")]
 FailedRead(AnyError),
```

`io::Error` and `serde_json::Error` don't implement `std::error:Error`. Is there another way, besides making it:

``` rust
#[error("Failed to write to the accounts file: {0}")]
FailedWrite(String),

 #[error("Failed to read the accounts file: {0}")]
 FailedRead(String),
```

and format the actual error to `String` to when building `FailedWrite` and `FailedRead`?

Reply:
> io::Error and serde_json::Error don't implement std::error:Error

I've just checked the docs and both of them do :thinking:

Reply:
Oh, can you pointed me there? https://docs.serde.rs/serde_json/struct.Error.html doesn't show any std::error::Error implementation

Reply:
I can't directly link to `impl`s, so I've made a screenshot

![image](https://user-images.githubusercontent.com/26183680/78258332-5df25f80-74fb-11ea-9d8f-78295404d0f9.png)

Reply:
The problem with ambiguity again... Anyway, now that I have implemented `from` using `thiserror` and creating these new error variants is much cleaner I'd leave it as is. If you don't oppose, I will merge.

Reply:
You're right, it DOES improve the code down the line. It turned out to be a good design.